### ISSUE TITLE
pycdf fails with non-unicode characters in CDF attribute

### ISSUE THREAD

**Issue by drsteve** [spacepy/spacepy#555]
Loading ARASE ephemeris CDFs with `spacepy.datamodel.fromCDF` fails with a `UnicodeDecodeError`.
This arises because `fromCDF` is reading _everything_ at once, but the actual error is in `pycdf`.

Basically, there is a non-unicode byte sequence at the end of one of the metadata entries. There's no failsafe on conversion (e.g., a replace mode, or a skip-if-invalid), so the attribute read fails.

### Minimal example to reproduce issue:
```
wget https://ergsc.isee.nagoya-u.ac.jp/data/ergsc/satellite/erg/orb/def/2017/erg_orb_l2_20170908_v03.cdf
```
then in python
```
import spacepy.pycdf as cdf
data = cdf.CDF('erg_orb_l2_20170908_v03.cdf')
data.attrs
```

### Error message/Traceback:
```
Python 3.8.2 (default, Mar 26 2020, 15:53:00) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.26.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import spacepy.pycdf as cdf
This unreleased version of SpacePy is not supported by the SpacePy team.
/home/smorley/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/time.py:2365: UserWarning: Leapseconds may be out of date. Use spacepy.toolbox.update(leapsecs=True)
  warnings.warn('Leapseconds may be out of date.'

In [2]: data = cdf.CDF('erg_orb_l2_20170908_v03.cdf')

In [3]: data.attrs
Out[3]: ---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
~/miniconda3/envs/default/lib/python3.8/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/miniconda3/envs/default/lib/python3.8/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395 
    396             return _default_pprint(obj, self, cycle)

~/miniconda3/envs/default/lib/python3.8/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in __repr__(self)
   5580         @rtype: str
   5581         """
-> 5582         return '<' + self.__class__.__name__ + ':\n' + str(self) + '\n>'
   5583 
   5584     def __str__(self):

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in __str__(self)
   5592         """
   5593         if self._cdf_file._opened:
-> 5594             return '\n'.join([key + ': ' + (
   5595                 ('\n' + ' ' * (len(key) + 2)).join(
   5596                 [str(value[i]) + ' [' + lib.cdftypenames[value.type(i)] + ']'

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in <listcomp>(.0)
   5594             return '\n'.join([key + ': ' + (
   5595                 ('\n' + ' ' * (len(key) + 2)).join(
-> 5596                 [str(value[i]) + ' [' + lib.cdftypenames[value.type(i)] + ']'
   5597                  for i in range(value.max_idx() + 1) if value.has_entry(i)])
   5598                 if isinstance(value, Attr)

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in <listcomp>(.0)
   5594             return '\n'.join([key + ': ' + (
   5595                 ('\n' + ' ' * (len(key) + 2)).join(
-> 5596                 [str(value[i]) + ' [' + lib.cdftypenames[value.type(i)] + ']'
   5597                  for i in range(value.max_idx() + 1) if value.has_entry(i)])
   5598                 if isinstance(value, Attr)

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in __getitem__(self, key)
   4751         else:
   4752             if self.has_entry(key):
-> 4753                 return self._get_entry(key)
   4754             else:
   4755                 raise IndexError('list index ' + str(key) + ' out of range.')

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/pycdf/__init__.py in _get_entry(self, number)
   5230                 result = bytes(buff)
   5231             else: #Py3k, make unicode
-> 5232                 result = str(numpy.char.array(buff).decode())
   5233         else:
   5234             if not self._raw:

~/.local/lib/python3.8/site-packages/numpy/core/defchararray.py in decode(self, encoding, errors)
   2185 
   2186         """
-> 2187         return decode(self, encoding, errors)
   2188 
   2189     def encode(self, encoding=None, errors=None):

<__array_function__ internals> in decode(*args, **kwargs)

~/.local/lib/python3.8/site-packages/numpy/core/defchararray.py in decode(a, encoding, errors)
    555     """
    556     return _to_string_or_unicode_array(
--> 557         _vec_string(a, object_, 'decode', _clean_args(encoding, errors)))
    558 
    559 

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa1 in position 407: invalid start byte
```

### OS, Python version, and dependency version information:
Linux-4.19.128-microsoft-standard-x86_64-with-glibc2.10
sys.version_info(major=3, minor=8, micro=2, releaselevel='final', serial=0)
numpy=1.21.2
scipy=1.6.2
matplotlib=3.4.3
CDF=3.8.0.1

(note that this was my reproduction of someone else's bug report - all I know about their system is Python 3.9, Mac, SpacePy from github fairly recently)

### Version of SpacePy
HEAD of master from repo

**Comment by drsteve** [comment]
For additional info, the final bytes of the attribute it's trying to read appear to be '0xa1 0xb1'. I have no idea what encoding this is supposed to be, but using 'latin_1' or 'shift_jis' both allow the decode (even if they give what is probably garbage for the last couple of characters).

My initial inclination here is to raise a warning and I think there's a replace mode for decoding where invalid bytes get replaced by "?"... thoughts?

**Comment by jtniehof** [comment]
Likely to be fixed with #586 but not guaranteed; will check this use case in implementing that.

**Comment by jtniehof** [comment]
I can't get that gEntry to decode to anything reasonable with UTF-8, shift-JIS, or latin-1, but at least I can get the replacement characters with the fix for #586.
